File: integration-prototypes/csharp/opensearch-ha-rest/tests/OpenSearchHaRestService.Tests/OpenSearchHaTests.cs
using OpenSearchHaRestService;

namespace OpenSearchHaRestService.Tests;

public sealed class OpenSearchHaTests
{
    [Fact]
    public void ParseNodes_ReturnsAllNodes()
    {
        var nodes = OpenSearchHaRepository.ParseNodes("http://localhost:9200,http://localhost:9201,http://localhost:9202");
        Assert.Equal(3, nodes.Count);
    }

    [Fact]
    public async Task FetchValueAsync_UsesFailover()
    {
        var repository = new OpenSearchHaRepository();
        var calls = new List<string>();

        var value = await repository.FetchValueAsync(
            "prototype_docs",
            "1",
            (node, _index, _id) =>
            {
                calls.Add(node);
                if (node.EndsWith(":9200", StringComparison.Ordinal))
                {
                    throw new InvalidOperationException("node down");
                }

                return Task.FromResult<string?>("hello opensearch ha");
            });

        Assert.Equal("hello opensearch ha", value);
        Assert.True(calls.Count >= 2);
    }
}

File: integration-prototypes/java/opensearch-ha-rest/src/test/java/org/prototype/opensearch/ha/OpenSearchHaTests.java
package org.prototype.opensearch.ha;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.net.http.HttpClient;
import org.junit.jupiter.api.Test;

public final class OpenSearchHaTests {
    @Test
    void parseNodesCount() {
        assertEquals(3, OpenSearchHaRepository.parseNodes("http://localhost:9200,http://localhost:9201,http://localhost:9202").size());
    }

    @Test
    void fetchValueFailover() throws Exception {
        OpenSearchHaRepository repository = new OpenSearchHaRepository("http://localhost:9200,http://localhost:9201", HttpClient.newHttpClient()) {
            @Override
            protected String fetchFromNode(String nodeUrl, String index, String id) {
                if (nodeUrl.endsWith(":9200")) {
                    throw new IllegalStateException("node down");
                }
                return "hello opensearch ha";
            }
        };

        assertEquals("hello opensearch ha", repository.fetchValue("prototype_docs", "1"));
    }

    @Test
    void fetchValueAllNodesDown() {
        OpenSearchHaRepository repository = new OpenSearchHaRepository("http://localhost:9200", HttpClient.newHttpClient()) {
            @Override
            protected String fetchFromNode(String nodeUrl, String index, String id) {
                throw new IllegalStateException("node down");
            }
        };

        assertThrows(IllegalStateException.class, () -> repository.fetchValue("prototype_docs", "1"));
    }
}
